module7/Homework7-start/Homework7Tests/UserStoreTests.swift

```diff
@@ -0,0 +1,113 @@
+/// Copyright (c) 2024 Kodeco Inc.
+/// 
+/// Permission is hereby granted, free of charge, to any person obtaining a copy
+/// of this software and associated documentation files (the "Software"), to deal
+/// in the Software without restriction, including without limitation the rights
+/// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+/// copies of the Software, and to permit persons to whom the Software is
+/// furnished to do so, subject to the following conditions:
+/// 
+/// The above copyright notice and this permission notice shall be included in
+/// all copies or substantial portions of the Software.
+/// 
+/// Notwithstanding the foregoing, you may not use, copy, modify, merge, publish,
+/// distribute, sublicense, create a derivative work, and/or sell copies of the
+/// Software in any work that is designed, intended, or marketed for pedagogical or
+/// instructional purposes related to programming, coding, application development,
+/// or information technology.  Permission for such use, copying, modification,
+/// merger, publication, distribution, sublicensing, creation of derivative works,
+/// or sale is expressly withheld.
+/// 
+/// This project and source code may use libraries or frameworks that are
+/// released under various Open-Source licenses. Use of those libraries and
+/// frameworks are governed by their own individual licenses.
+///
+/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+/// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+/// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+/// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+/// THE SOFTWARE.
+
+import XCTest
+
+final class UserStoreTests: XCTestCase {
+  var userStore: UserStore!
+  let missingURL = URL(fileURLWithPath: "missingFile",
+                       relativeTo: FileManager.documentsDirectoryURL).appendingPathExtension("json")
+  
+  
+  override func setUpWithError() throws {
+    userStore = UserStore()
+  }
+  
+  override func tearDownWithError() throws {
+    if FileManager.default.fileExists(atPath: userStore.documentsJSONURL.path) {
```

Comment:
be mindful of the fact that the test engine will run tests in any order and will run tests concurrently, you wouldn't want your teardown of one test to cause a different test to fail.

Reply:
Great catch. Concurrency errors like these can cause lost hours debugging flaky tests in a production setting at scale. 

It is interesting what's the best way to handle this? One option (which I did) is to move the clean up close to the test that actually needs it. But this will just minimize the side-effects and if other tests are added that need to access or mutate the file at this path. I wonder if there's an XCode setting to disable parallelism for certain TestCases or modules?

Reply:
What I see a lot is moving the setup/teardown code into the test and unless the filename is set, makeing sure that the filename is unique to this test. Another way is to write mockers. This is a little more involved, but especially when you get to networking tests where the network may not be available, or may be so slow that your test suites take forever. The general idea around a mocker is that you make some class in your code that handles the networking or at least is in the chain between your business code and the network. Then in the test, you subclass that with something that just returns a known value for the test.